src/Microsoft.AspNet.Server.Kestrel/Http/Frame.cs

```diff
@@ -218,33 +288,37 @@ private void FireOnCompleted()
             }
         }
 
-        private async Task ExecuteAsync()
+        public void Flush()
         {
-            Exception error = null;
-            try
-            {
-                await Application.Invoke(this).ConfigureAwait(false);
+            ProduceStart(immediate: false);
+            SocketOutput.Write(_emptyData, immediate: true);
+        }
+
+        public Task FlushAsync(CancellationToken cancellationToken)
+        {
+            ProduceStart(immediate: false);
+            return SocketOutput.WriteAsync(_emptyData, immediate: true);
+        }
+
+        public void Write(ArraySegment<byte> data)
+        {
+            ProduceStart(immediate: false);
 
-                // Trigger FireOnStarting if ProduceStart hasn't been called yet.
-                // We call it here, so it can go through our normal error handling
-                // and respond with a 500 if an OnStarting callback throws.
-                if (!_responseStarted)
+            if (_autoChunk)
+            {
+                if (data.Count == 0)
```

Comment:
I know this is the way it was before, but would it make sense to exit early even if we aren't chunking for zero length writes?